<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Gather summary statistics for exposures — gather_sumstats • FAMR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Gather summary statistics for exposures — gather_sumstats"><meta name="description" content="Given directories containing summary statistics and LD matrices for one or
more independent loci, reads in this data and merges it with data for the
outcome trait, if applicable. Also performs LD clumping to reduce
computational cost."><meta property="og:description" content="Given directories containing summary statistics and LD matrices for one or
more independent loci, reads in this data and merges it with data for the
outcome trait, if applicable. Also performs LD clumping to reduce
computational cost."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FAMR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.8</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Gather summary statistics for exposures</h1>

      <div class="d-none name"><code>gather_sumstats.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Given directories containing summary statistics and LD matrices for one or
more independent loci, reads in this data and merges it with data for the
outcome trait, if applicable. Also performs LD clumping to reduce
computational cost.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gather_sumstats</span><span class="op">(</span></span>
<span>  <span class="va">in_dir</span>,</span>
<span>  <span class="va">ld_dir</span>,</span>
<span>  y_gwas <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  oracle_mode <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  prune_thresh <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  fa_prune_thresh <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-in-dir">in_dir<a class="anchor" aria-label="anchor" href="#arg-in-dir"></a></dt>
<dd><p>A directory with input rds files containing summary statistics
for the exposures, with one file per locus. Each file should be a list object
with three matrices with SNPs as rows and exposure traits as columns. These
are "betas" (effect size estimates of SNPs on exposures), "stderrs"
(standard errors of the betas), and "Z" (the z-scores, i.e. betas/stderrs).
Optionally can also include corresponding vectors betas_y, stderrs_y, and
zscores_y for the outcome trait ("y"); alternatively, those can be read in
separately via the read_y_gwas function.
Finally, it has a vector 'pos' with the SNP names.</p></dd>


<dt id="arg-ld-dir">ld_dir<a class="anchor" aria-label="anchor" href="#arg-ld-dir"></a></dt>
<dd><p>A directory with input rds files containing LD matrices
corresponding to the SNPs in the summary statistics files, with one file
per locus. These should simply be numerical nSNP-by-nSNP matrices.</p></dd>


<dt id="arg-y-gwas">y_gwas<a class="anchor" aria-label="anchor" href="#arg-y-gwas"></a></dt>
<dd><p>A list with four items: names_y, betas_y, stderrs_y, and
zscores_y, giving the IDs, estimated effect sizes, standard errors, and
z-scores of the SNPs on the outcome trait, respectively. Usually produced
by the read_y_gwas function. If data for the outcome is included in the files
in in_dir, leave this blank (default).</p></dd>


<dt id="arg-oracle-mode">oracle_mode<a class="anchor" aria-label="anchor" href="#arg-oracle-mode"></a></dt>
<dd><p>DEPRECATED. Used for internal testing purposes.</p></dd>


<dt id="arg-prune-thresh">prune_thresh<a class="anchor" aria-label="anchor" href="#arg-prune-thresh"></a></dt>
<dd><p>LD clumping threshold; a numeric value between 0 and 1.
Default is 0.5.</p></dd>


<dt id="arg-fa-prune-thresh">fa_prune_thresh<a class="anchor" aria-label="anchor" href="#arg-fa-prune-thresh"></a></dt>
<dd><p>Stricter LD clumping threshold used to generate a set
of SNPs to be input to factor analysis. Default: 0.1.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list object with the following elements:</p>
<dl><dt>betas</dt>
<dd><p>A matrix of estimated effect sizes of SNPs on exposures, with
rows corresponding to SNPs and columns corresponding to exposures.</p></dd>


<dt>weights</dt>
<dd><p>Currently this is the same as betas; later used to store
learned weights of SNPs on exposures and factors.</p></dd>


<dt>stderrs</dt>
<dd><p>A matrix of standard errors for the betas.</p></dd>


<dt>Z</dt>
<dd><p>A matrix of z-scores for the exposures, equal to betas divided by
stderrs.</p></dd>


<dt>betas_y</dt>
<dd><p>A vector of estimated effect sizes of SNPs on the outcome.</p></dd>


<dt>stderrs_y</dt>
<dd><p>A vector of standard errors for betas_y.</p></dd>


<dt>zscores_y</dt>
<dd><p>A vector of z-scores of SNPs on the outcome, equal to
betas_y divided by stderrs_y.</p></dd>


<dt>pos</dt>
<dd><p>A vector of SNP names.</p></dd>


<dt>locus_idx</dt>
<dd><p>A vector matching SNPs to which locus they come from,
indexed in the order the loci are read from in_dir.</p></dd>


<dt>lambda</dt>
<dd><p>A matrix of lambda values learned by learn_lambda for each
exposure in each locus, where each row of the matrix corresponds to a locus
and the columns are the exposures.</p></dd>


<dt>ss_for_fa</dt>
<dd><p>A list object with the "betas", "stderrs", "Z", "weights",
and "pos" fields, but only for the SNPs to be input to factor analysis.</p></dd>


</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nathan LaPierre.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

