<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Sum of Single Effects (SuSiE) Regression using summary statistics — susie_rss • FAMR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Sum of Single Effects (SuSiE) Regression using summary statistics — susie_rss"><meta name="description" content="susie_rss performs sum of single-effect linear
regression with z scores; all posterior calculations are for
z-scores. This function fits the regression model \(z = \sum_l
  R*b_l + e\), where e is \(N(0,R)\) and \(\sum_l b_l\) is a
p-vector of effects to be estimated. The required summary data are
the p by p correlation matrix, R, and the p-vector z."><meta property="og:description" content="susie_rss performs sum of single-effect linear
regression with z scores; all posterior calculations are for
z-scores. This function fits the regression model \(z = \sum_l
  R*b_l + e\), where e is \(N(0,R)\) and \(\sum_l b_l\) is a
p-vector of effects to be estimated. The required summary data are
the p by p correlation matrix, R, and the p-vector z."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FAMR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.13</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Getting_Started.html">Getting_Started</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/nlapier2/FAMR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Sum of Single Effects (SuSiE) Regression using summary statistics</h1>
      <small class="dont-index">Source: <a href="https://github.com/nlapier2/FAMR/blob/main/R/susie_ctwas_all.R" class="external-link"><code>R/susie_ctwas_all.R</code></a></small>
      <div class="d-none name"><code>susie_rss.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>susie_rss</code> performs sum of single-effect linear
regression with z scores; all posterior calculations are for
z-scores. This function fits the regression model \(z = \sum_l
  R*b_l + e\), where e is \(N(0,R)\) and \(\sum_l b_l\) is a
p-vector of effects to be estimated. The required summary data are
the p by p correlation matrix, <code>R</code>, and the p-vector <code>z</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">susie_rss</span><span class="op">(</span></span>
<span>  <span class="va">z</span>,</span>
<span>  <span class="va">R</span>,</span>
<span>  maf <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  maf_thresh <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  z_ld_weight <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  L <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  prior_variance <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  residual_variance <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  prior_weights <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  null_weight <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  estimate_residual_variance <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  estimate_prior_variance <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  estimate_prior_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"optim"</span>, <span class="st">"EM"</span>, <span class="st">"simple"</span><span class="op">)</span>,</span>
<span>  check_null_threshold <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  prior_tol <span class="op">=</span> <span class="fl">1e-09</span>,</span>
<span>  max_iter <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  s_init <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  intercept_value <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  coverage <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  min_abs_corr <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  tol <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  track_fit <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  check_R <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  r_tol <span class="op">=</span> <span class="fl">1e-08</span>,</span>
<span>  refine <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-z">z<a class="anchor" aria-label="anchor" href="#arg-z"></a></dt>
<dd><p>A p-vector of z scores.</p></dd>


<dt id="arg-r">R<a class="anchor" aria-label="anchor" href="#arg-r"></a></dt>
<dd><p>A p by p symmetric, positive semidefinite correlation
matrix.</p></dd>


<dt id="arg-maf">maf<a class="anchor" aria-label="anchor" href="#arg-maf"></a></dt>
<dd><p>Minor allele frequency; to be used along with
<code>maf_thresh</code> to filter input summary statistics.</p></dd>


<dt id="arg-maf-thresh">maf_thresh<a class="anchor" aria-label="anchor" href="#arg-maf-thresh"></a></dt>
<dd><p>Variants having a minor allele frequency smaller
than this threshold are not used.</p></dd>


<dt id="arg-z-ld-weight">z_ld_weight<a class="anchor" aria-label="anchor" href="#arg-z-ld-weight"></a></dt>
<dd><p>This feature is not recommended. The weights
assigned to the z scores in the LD matrix. If <code>z_ld_weight &gt;
  0</code>, the LD matrix used in the model is <code>cov2cor((1-w)*R +
  w*tcrossprod(z))</code>, where <code>w = z_ld_weight</code>.</p></dd>


<dt id="arg-l">L<a class="anchor" aria-label="anchor" href="#arg-l"></a></dt>
<dd><p>Number of components (nonzero coefficients) in the susie
regression model. If L is larger than the number of covariates, p,
L is set to p.</p></dd>


<dt id="arg-prior-variance">prior_variance<a class="anchor" aria-label="anchor" href="#arg-prior-variance"></a></dt>
<dd><p>The prior variance. It is either a scalar or
a vector of length L.</p></dd>


<dt id="arg-residual-variance">residual_variance<a class="anchor" aria-label="anchor" href="#arg-residual-variance"></a></dt>
<dd><p>Variance of the residual.
If it is not specified, we set it to 1.</p></dd>


<dt id="arg-prior-weights">prior_weights<a class="anchor" aria-label="anchor" href="#arg-prior-weights"></a></dt>
<dd><p>A vector of length p, in which each entry
gives the prior probability that SNP j has non-zero effect.</p></dd>


<dt id="arg-null-weight">null_weight<a class="anchor" aria-label="anchor" href="#arg-null-weight"></a></dt>
<dd><p>Prior probability of no effect (a number between
0 and 1, and cannot be exactly 1).</p></dd>


<dt id="arg-estimate-residual-variance">estimate_residual_variance<a class="anchor" aria-label="anchor" href="#arg-estimate-residual-variance"></a></dt>
<dd><p>The residual variance is
fixed to the value supplied by <code>residual_variance</code>. We don't
estimate residual variance in susie_rss.</p></dd>


<dt id="arg-estimate-prior-variance">estimate_prior_variance<a class="anchor" aria-label="anchor" href="#arg-estimate-prior-variance"></a></dt>
<dd><p>If <code>estimate_prior_variance =
  TRUE</code>, the prior variance is estimated (this is a separate
parameter for each of the L effects). If provided,
<code>prior_variance</code> is then used as an initial value for
the optimization. When <code>estimate_prior_variance = FALSE</code>, the
prior variance for each of the L effects is determined by the
value supplied to <code>prior_variance</code>.</p></dd>


<dt id="arg-estimate-prior-method">estimate_prior_method<a class="anchor" aria-label="anchor" href="#arg-estimate-prior-method"></a></dt>
<dd><p>The method used for estimating prior
variance. When <code>estimate_prior_method = "simple"</code> is used, the
likelihood at the specified prior variance is compared to the
likelihood at a variance of zero, and the setting with the larger
likelihood is retained.</p></dd>


<dt id="arg-check-null-threshold">check_null_threshold<a class="anchor" aria-label="anchor" href="#arg-check-null-threshold"></a></dt>
<dd><p>When the prior variance is estimated,
compare the estimate with the null, and set the prior variance to
zero unless the log-likelihood using the estimate is larger by this
threshold amount. For example, if you set
<code>check_null_threshold = 0.1</code>, this will "nudge" the estimate
towards zero when the difference in log-likelihoods is small. A
note of caution that setting this to a value greater than zero may
lead the IBSS fitting procedure to occasionally decrease the ELBO.</p></dd>


<dt id="arg-prior-tol">prior_tol<a class="anchor" aria-label="anchor" href="#arg-prior-tol"></a></dt>
<dd><p>When the prior variance is estimated, compare the
estimated value to <code>prior_tol</code> at the end of the computation,
and exclude a single effect from PIP computation if the estimated
prior variance is smaller than this tolerance value.</p></dd>


<dt id="arg-max-iter">max_iter<a class="anchor" aria-label="anchor" href="#arg-max-iter"></a></dt>
<dd><p>Maximum number of IBSS iterations to perform.</p></dd>


<dt id="arg-s-init">s_init<a class="anchor" aria-label="anchor" href="#arg-s-init"></a></dt>
<dd><p>A previous susie fit with which to initialize.</p></dd>


<dt id="arg-intercept-value">intercept_value<a class="anchor" aria-label="anchor" href="#arg-intercept-value"></a></dt>
<dd><p>The intercept. (The intercept cannot be
estimated from centered summary data.) This setting will be used by
<code>coef</code> to assign an intercept value, mainly for consistency
with <code>susie</code>. Set to <code>NULL</code> if you want <code>coef</code> not
to include an intercept term (and so only a p-vector is returned).</p></dd>


<dt id="arg-coverage">coverage<a class="anchor" aria-label="anchor" href="#arg-coverage"></a></dt>
<dd><p>A number between 0 and 1 specifying the
“coverage” of the estimated confidence sets.</p></dd>


<dt id="arg-min-abs-corr">min_abs_corr<a class="anchor" aria-label="anchor" href="#arg-min-abs-corr"></a></dt>
<dd><p>Minimum absolute correlation allowed in a
credible set. The default, 0.5, corresponds to a squared
correlation of 0.25, which is a commonly used threshold for
genotype data in genetic studies.</p></dd>


<dt id="arg-tol">tol<a class="anchor" aria-label="anchor" href="#arg-tol"></a></dt>
<dd><p>A small, non-negative number specifying the convergence
tolerance for the IBSS fitting procedure. The fitting procedure
will halt when the difference in the variational lower bound, or
“ELBO” (the objective function to be maximized), is
less than <code>tol</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>If <code>verbose = TRUE</code>, the algorithm's progress,
and a summary of the optimization settings, are printed to the
console.</p></dd>


<dt id="arg-track-fit">track_fit<a class="anchor" aria-label="anchor" href="#arg-track-fit"></a></dt>
<dd><p>If <code>track_fit = TRUE</code>, <code>trace</code>
is also returned containing detailed information about the
estimates at each iteration of the IBSS fitting procedure.</p></dd>


<dt id="arg-check-r">check_R<a class="anchor" aria-label="anchor" href="#arg-check-r"></a></dt>
<dd><p>If <code>check_R = TRUE</code>, check that <code>R</code> is
positive semidefinite.</p></dd>


<dt id="arg-r-tol">r_tol<a class="anchor" aria-label="anchor" href="#arg-r-tol"></a></dt>
<dd><p>Tolerance level for eigenvalue check of positive
semidefinite matrix of R.</p></dd>


<dt id="arg-refine">refine<a class="anchor" aria-label="anchor" href="#arg-refine"></a></dt>
<dd><p>If <code>refine = TRUE</code>, we use a procedure to help
SuSiE get out of local optimum.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>"susie"</code> object with some or all of the following
elements:</p>
<dl><dt>alpha</dt>
<dd><p>An L by p matrix of posterior inclusion probabilites.</p></dd>


<dt>mu</dt>
<dd><p>An L by p matrix of posterior means, conditional on
inclusion.</p></dd>


<dt>mu2</dt>
<dd><p>An L by p matrix of posterior second moments,
conditional on inclusion.</p></dd>


<dt>lbf</dt>
<dd><p>log-Bayes Factor for each single effect.</p></dd>


<dt>lbf_variable</dt>
<dd><p>log-Bayes Factor for each variable and single effect.</p></dd>


<dt>intercept</dt>
<dd><p>Fixed Intercept.</p></dd>


<dt>sigma2</dt>
<dd><p>Fixed Residual variance.</p></dd>


<dt>V</dt>
<dd><p>Prior variance of the non-zero elements of b, equal to
<code>scaled_prior_variance * var(Y)</code>.</p></dd>


<dt>elbo</dt>
<dd><p>The value of the variational lower bound, or
“ELBO” (objective function to be maximized), achieved at
each iteration of the IBSS fitting procedure.</p></dd>


<dt>fitted</dt>
<dd><p>Vector of length n containing the fitted values of
the outcome.</p></dd>


<dt>sets</dt>
<dd><p>Credible sets estimated from model fit; see
<code><a href="susie_get_methods.html">susie_get_cs</a></code> for details.</p></dd>


<dt>pip</dt>
<dd><p>A vector of length p giving the (marginal) posterior
inclusion probabilities for all p covariates.</p></dd>


<dt>niter</dt>
<dd><p>Number of IBSS iterations that were performed.</p></dd>


<dt>converged</dt>
<dd><p><code>TRUE</code> or <code>FALSE</code> indicating whether
the IBSS converged to a solution within the chosen tolerance
level.</p></dd>


<dt>Rr</dt>
<dd><p>An p-vector of <code>t(X)</code> times fitted values, <code>X
  %*% colSums(alpha*mu)</code>.</p></dd>


</dl><p>@keywords internal</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nathan LaPierre, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

